### PR TITLE
SLCORE-826 Let clients provide user-configured local file exclusions

### PR DESCRIPTION
**Pull request by serhat-yenican-sonarsource**
# For SonarSourcers:

validated in IntelliJ. IntelliJ test branch: poc/serhat/client-file-exclusions

- [x] Prefix the commit message with the ticket number, i.e. `SLCORE-XXXX` (or use `NO-JIRA` for a trivial change not tracked by a ticket)
- [ ] When changing an API:
    - [x] Explain in the JavaDoc the purpose of the new API
    - [x] Document the change in [API_CHANGES.md](https://github.com/SonarSource/sonarlint-core/blob/master/API_CHANGES.md)
    - [ ] If the change breaks the current API, explicitly communicate those to the impacted consumers prior to merging (eg. IDE squad)
- [x] Make sure the tests adhere to the convention:
    - [ ] All test method names should use `snake_case`, for example: `test_validate_input`.
- [ ] Make sure checks are green: build passes, Quality Gate is green

# For external contributors:

In addition to the above, please review our [contribution guidelines](https://github.com/SonarSource/sonarlint-core/blob/master/docs/contributing.md) and ensure your pull request adheres to the following guidelines:

- [ ] Please explain your motives to contribute this change: what problem you are trying to fix, what improvement you are trying to make
- [ ] Use the following formatting style: [SonarSource/sonar-developer-toolset](https://github.com/SonarSource/sonar-developer-toolset#code-style)
- [ ] Provide a unit test for any code you changed

### CODE REVIEW

File: medium-tests/src/test/java/mediumtest/file/ClientFileExclusionsMediumTests.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+      .containsOnlyKeys(fileUri)
+      .hasEntrySatisfying(fileUri, issues -> {
+        assertThat(issues)
+          .extracting(RaisedIssueDto::getPrimaryMessage)
+          .containsExactly("Replace \"pom.version\" with \"project.version\".");
+      });
+  }
+
+  @Test
+  void it_should_not_exclude_client_defined_file_exclusion_in_standalone_mode(@TempDir Path baseDir) {
```
Comment (kirill-knize-sonarsource) [inline medium-tests/src/test/java/mediumtest/file/ClientFileExclusionsMediumTests.java:121]: in connected mode?

File: rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/analysis/GetFileExclusionsResponse.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ */
+package org.sonarsource.sonarlint.core.rpc.protocol.client.analysis;
+
+import java.util.Set;
+import javax.annotation.Nullable;
+
+public class GetFileExclusionsResponse {
+  private final Set<String> fileExclusionPatterns;
+
+  public GetFileExclusionsResponse(@Nullable Set<String> fileExclusionPatterns) {
```
Comment (kirill-knize-sonarsource) [inline rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/analysis/GetFileExclusionsResponse.java:28]: Why nullable? Shouldn't it be an empty set then? And if it's nullable, the getter should be marked as @CheckForNull as well.

File: API_CHANGES.md [hunk 3]
```diff
@@ -10,6 +10,10 @@ None :)
 * Add a method to the `org.sonarsource.sonarlint.core.rpc.protocol.backend.analysis.AnalysisRpcService` to let the client notify the backend with user defined analysis properties
   * `org.sonarsource.sonarlint.core.rpc.protocol.backend.analysis.AnalysisRpcService#didSetUserAnalysisProperties` to set user defined properties
 * For analysis, both user-defined and inferred properties will be merged. If the same property is inferred by the client and provided by the user - the inferred value will be used for analysis.
+* Add a method to `org.sonarsource.sonarlint.core.rpc.protocol.SonarLintRpcClient` to allow backend to request client-defined file
+  exclusions from the
```
Comment (kirill-knize-sonarsource) [inline API_CHANGES.md:14]: Too many new lines. It all should be one line on line 13, I think.

File: rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/analysis/GetFileExclusionsResponse.java (hunk 2 above)
Comment (serhat-yenican-sonarsource) [reply rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/analysis/GetFileExclusionsResponse.java:28]: Exactly! It's a copy-paste issue

**Review by Knize — APPROVED** [review]
LGTM!

**Review by kirill-knize-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**